command/agent/command.go

```diff
@@ -427,10 +432,130 @@ func (c *Command) startAgent(config *Config, agent *Agent,
 	if config.Discover != "" {
 		c.Ui.Info(fmt.Sprintf("  mDNS cluster: %s", config.Discover))
 	}
+
 	return ipc
 }
 
 // startupJoin is invoked to handle any joins specified to take place at start time
+func (c *Command) startupJoinSRV(config *Config, agent *Agent) error {
+	if len(config.JoinSRV) == 0 {
+		return nil
+	}
+
+	c.Ui.Output(fmt.Sprintf("Joining cluster via SRV...(replay: %v)", config.ReplayOnJoin))
+	n, err := c.joinSRV(agent, config.ReplayOnJoin, config.JoinSRV)
+	if err != nil {
+		return err
+	}
+
+	if n == 0 {
+		return errors.New("Failed to join any hosts via SRV")
+	}
+
+	c.Ui.Info(fmt.Sprintf("Join completed. Synced with %d initial agents", n))
+	return nil
+
+}
+
+// retryJoinSRV is invoked to handle joins with retries. This runs until at least a
+// single successful join or RetryMaxAttempts is reached
+func (c *Command) retryJoinSRV(config *Config, agent *Agent, errCh chan struct{}) {
+	// Quit fast if there is no nodes to join
+	if len(config.RetryJoinSRV) == 0 {
+		return
+	}
+
+	// Track the number of join attempts
+	attempt := 0
+	for {
+		// Try to perform the join
+		n, err := c.joinSRV(agent, config.ReplayOnJoin, config.RetryJoinSRV)
+
+		if err != nil {
+			c.logger.Printf("[ERR] agent: Failed to join via SRV: %v", err)
+		}
+
+		if err == nil && n > 0 {
+			c.logger.Printf("[INFO] agent: Join completed. Synced with %d initial agents", n)
```

Comment:
@slackpad the bastions find each other through the SRV record once it's registered, then all other nodes find the bastions.

The continuous polling is useful for any other nodes (private or public doesn't really matter) to discover the cluster. The only case where I think it's totally necessary is when both bastions go away, which would result in the original cluster becoming 'orphaned'.

A workaround would be to have the nodes register themselves as part of the SRV record, but I'd prefer to not have the nodes updating the records and to just keep them in terraform as that's a lot more predictable.

Or we could just stick with the crontab approach, but I think that the continuous polling to prevent orphaned clusters is the simplest, most reliable approach. Under the current implementation, it never has to try a join if the cluster is already fully joined.